src/frontend/AppContainerWrapper.tsx

```diff
@@ -108,45 +136,19 @@ const AppContainerWrapper = () => {
     []
   );
 
```

Comment:
This `updateRouteBasedAppState` should be derived from the routeName within the render, rather than through a callback and setting state. This is going to cause a lot of re-renders for each navigation change.

Reply:
e.g. store the routeName in state and derive everything else from that.

Reply:
This is @achou11's architecture. Im curious what you both think. This app container needs the route name, but the route name is created inside the children of the appcontainer. So we could create another context, and pass the route name to that context....but then we are adding another context (which is not necessarily a problem). Is that what you had in mind @gmaclennan ?

Reply:
yeah gregor's right in that the state being set in that function should be probably be derived. Not exactly sure what my reasoning here was at the time - probably just too enamored with using effects 😅 but yeah, storing the route name in state and converting the states it updates to be derived should work?

Reply:
@achou11 just for clarification, we would need to do this through a context, correct?

Reply:
I dont think so? the route name state would live in this component and the calculated values are only needed here, right? i haven't looked at the updated implementation from your PR so apologies if i'm missing something

Reply:
yep!

Reply:
This is getting off topic (because of me) because this is not what this PR is about, but it's a chance to cleanup code. Does this definitely need to sit outside (e.g. a parent of) the AppContainer? Is there a way for practice mode and invite modal logic to be defined inside AppContainer, where routeName would be available as a prop / in context? That would avoid the slightly hairy logic of hoisting a navigation prop to the parent (which is essentially what this is doing). I don't think this needs to be resolved in this PR, but since this is my first proper review of this code I wanted to share thoughts on it.

Reply:
Unfortunately not. So the `AppContainer` is of type `Stack.Navigator`. `Stack.Navigator` can only have `Stack` Components as its children. As we soon as we put a non stack item(practice mode), it throws a run time error. We could maybe put the invite modal logic though.... @achou11 you designed that originally, what do you think?

Reply:
i haven't fully familiarized with the new navigation api but my underinformed understanding is that since practice mode is fundamentally a modification of the original stack i.e. same routes but with some UI adjustments and more limited access to certain screens, it may not make sense to create its own navigator for it and nest it due to potential duplication of screens. however, i do see the point of doing something like that to avoid having to account for some global state of "is in practice mode" for various screens, and it would make it easier to create the separation between modes from a development perspective i.e. a directory of screens that assume "normal" mode and another for "practice mode", where each can be built with the assumption of being under the correct navigation namespace/stack (e.g. `app/main/...` `app/practice/...`). only potential issues here may be 1) duplication and 2) sharing of screens (maybe this is not an issue though if properly architected). @ErikSin do you have a general sense of how many screens are explicitly affected by practice mode? if it's not too many, do you think creating a devoted stack navigator for it would be feasible? (maybe as a follow up PR)

Regarding the invite modal, i think that could be it's own navigation stack if it isn't already e.g. `app/invite/...`, so that there's less need for a global state to represent that. cant fully remember but are there things we disable that are unrelated to the invite flow when when we enter that flow?

Reply:
> Stack.Navigator can only have Stack Components as its children

also as a note, you should be able to nest different kinds of navigators within each other https://reactnavigation.org/docs/nesting-navigators/. not sure if that comment was referring to something else though

Reply:
Practice mode touches every page, so creating a new stack for it would just duplicate everything, so really not feasible. And yes, different navigators can be nested, but yeah not feasible for just a ui component like practice mode.

In regards to the invite modal....I was hoping you would remember, because I don't quite remember why we designed it the way we did 😬

Reply:
reading through Gregor's comment again, realizing that he's referring to something slightly different than what i was expanding on. i suppose something like this would be feasible, although i'm not entirely sure if there are downsides to this approach:

```jsx
const PseudoAppStack = createStackNavigator()
const AppStack = createStackNavigator()

const App = ({route, navigation}) => {
  // Use navigation and route params to do stuff related to invites, practice mode, etc
  return (
    <AppStack.Navigator>
      <AppStack.Screen .. />
    <AppStack.Navigator
  )
}

const AppContainer = () => {
  return (
    <PseudoAppContainer.Stack>
      <PseudoAppContainer.Screen 
        name="App"
        component={AppStack}
      />
    </PseudoAppContainer.Stack>
  )
}
```

Reply:
> Practice mode touches every page, so creating a new stack for it would just duplicate everything, so really not feasible

yeah sorry i meant **how many screens do something different because the app is in practice mode?** I understand that the practice mode UI is on most screens, but do all of the screens adjust internal logic because of the different mode? e.g. disabling or omitting certain UI or actions

Reply:
but yeah i think it probably isn't reasonable to go that route of a practice mode stack vs just representing it in state. see https://github.com/digidem/mapeo-mobile/pull/992#discussion_r913971557 for a more pertinent response to gregor's question

Reply:
For visibility, its does not actually work for the invite modal logic: re: [this comment](https://github.com/digidem/mapeo-mobile/pull/992#discussion_r915255574)

Reply:
just adding this snippet from a conversation with Erik that explains this rather convoluted setup right now. i had thought maybe using a [ref](https://reactnavigation.org/docs/navigation-container#methods-on-the-ref) could potentially solve the two problems at hand, but it only solves one:

> there are two problems that we’re trying to solve in an ideal situation:
>
> 1. we need to disable being able to open the invite modal flow depending on the screen in the root stack we’re in
> 2. we need to update the practice mode UI based on the route because it’s slightly different for certain screens
>
> 1 would be feasible to fix if we used the ref approach since we’re listening on an event listener for invites that are received. with the ref, we could determine the route the app is in when an processing an invite, and if the route is okay to show the invite modal, show it. in other words, the source of truth for our action is not the route, but the invite event
>
> 2 is harder and doesn’t benefit from the ref approach because the source of truth for our action is the route changing. using a ref wouldn’t cause state updates and thus rerenders that the practice mode UI depends on